src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs

```diff
@@ -137,18 +140,32 @@ static void CacheMethod(Type handler, Type messageType, Type interfaceGenericTyp
         static IEnumerable<Type> GetMessageTypesBeingHandledBy(Type type)
         {
             return (from t in type.GetInterfaces()
-                where t.IsGenericType
-                let potentialMessageType = t.GetGenericArguments()[0]
-                where
-                    typeof(IHandleMessages<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t) ||
-                    typeof(IHandleTimeouts<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t)
-                select potentialMessageType)
+                    where t.IsGenericType
+                    let potentialMessageType = t.GetGenericArguments()[0]
+                    where
+                        typeof(IHandleMessages<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t) ||
+                        typeof(IHandleTimeouts<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t)
+                    select potentialMessageType)
                 .Distinct()
                 .ToList();
         }
+        void ValidateHandlerType(Type handlerType)
+        {
+            var propertyTypes = handlerType.GetProperties().Select(p => p.PropertyType).ToList();
+            var ctorArguments = handlerType.GetConstructors()
+             .SelectMany(ctor => ctor.GetParameters().Select(p => p.ParameterType))
+             .ToList();
+
+            var dependencies = propertyTypes.Concat(ctorArguments).ToList();
+
+            if (dependencies.Any(t => typeof(IMessageSession).IsAssignableFrom(t)))
+            {
+                throw new Exception($"IMessageSession/IEndpointInstance shouldn't be injected into message handlers and Sagas. Use the context parameter on the `{handlerType.Name}.Handle` method to send messages.");
```

Comment:
@andreasohlund - Looks good. Minor tweaks.  How about:

Interfaces IMessageSession or IEndpointInstance should not be registered in the container to enable sending or publishing messages from within sagas or message handlers.  Instead, use the context parameter on the `{handlerType.Name}.Handle` method to send or publish messages.